src/transformers/models/funnel/modeling_tf_funnel.py

```diff
@@ -367,25 +369,24 @@ def pool_tensor(self, tensor, mode="mean", stride=2):
 
         return tf.squeeze(tensor, 2) if ndim == 2 else tensor
 
-    def pre_attention_pooling(self, output, attention_inputs):
+    def pre_attention_pooling(self, attention_inputs):
         """ Pool `output` and the proper parts of `attention_inputs` before the attention layer. """
         position_embeds, token_type_mat, attention_mask, cls_mask = attention_inputs
         if self.pool_q_only:
             if self.attention_type == "factorized":
                 position_embeds = self.stride_pool(position_embeds[:2], 0) + position_embeds[2:]
             token_type_mat = self.stride_pool(token_type_mat, 1)
             cls_mask = self.stride_pool(cls_mask, 0)
-            output = self.pool_tensor(output, mode=self.pooling_type)
         else:
             self.pooling_mult *= 2
             if self.attention_type == "factorized":
                 position_embeds = self.stride_pool(position_embeds, 0)
             token_type_mat = self.stride_pool(token_type_mat, [1, 2])
             cls_mask = self.stride_pool(cls_mask, [1, 2])
             attention_mask = self.pool_tensor(attention_mask, mode="min")
-            output = self.pool_tensor(output, mode=self.pooling_type)
+
         attention_inputs = (position_embeds, token_type_mat, attention_mask, cls_mask)
-        return output, attention_inputs
+        return attention_inputs
```

Comment:
Why change this function input and return? The tuple always has the same length so there is no reason for it to be incompatible with graph mode, no? The `output` can be pooled outside of the test to avoid the repetition of the same line of code, but otherwise, I'd prefer to keep the same logic as the PT implementation.

Reply:
I understand but this is not graph compliant. The reason is because `pooled_hidden` is not defined outside the `if pooling_flag` line 655. So to fix this I had to export the `pool_tensor` call outside the `if`.

Reply:
If `pooled_hidden` cannot be set outside if has to be deleted. Basically, either it has a value in both branches (same shape + same dtype) either it should not be set anywhere. Any idea what could be a "dummy" value?

Reply:
It's not used if `pooling_flag` is not used, so a tensor with a 0 can be a good dummy value (if it needs the same shape, a tensor of the right shape).

Reply:
Perfect! Doing the update.

Reply:
I don't succeed to guess the shape of `pooled_hidden` without calling `pool_tensor`. Any idea how to do this?

Reply:
`pooled_hidden = tf.zeros(shape_list(hidden))` seems to work perfectly fine 👍

Reply:
Yes, it should have the same shape as the hidden state if no pooling is done (so on the else side).